Warn user if Prisma needs to be re-generated because of prisma.schema changes

shayneczyzewski opened the pull request:
# Description

Currently, we do not warn users if Prisma needs to be re-generated via `wasp db migrate-dev` when they change their entities or db blocks. These changes will update `schema.prisma`, but if they do not apply it, it will not take effect.

This PR does two things:
- Every time `wasp db migrate-dev` is run, we take the SHA-256 checksum of the current `schema.prisma` file we are migrating against and store it as `schema.prisma.wasp-checksum` in the same .wasp/out/db dir. This means it is only local to the developer and serves as some sort of proof of prior local migration run.
- As part of Generation, we diff the `schema.prisma.wasp-checksum` file against the checksum of the current `schema.prisma` file. If they differ, we know the user has an unreflected update and they need to run `wasp db migrate-dev`. If they have no `schema.prisma.wasp-checksum` file but have entities, we also warn them to migrate.
  - Note: These checks run every compile, so not only `start` but `watch` too, so we can display an error to the console if they edit while still running.

Fixes #214 

## Type of change

Please select the option(s) that is more relevant.

- [x] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

Martinsos commented:
@shayneczyzewski looking forward to having this merged, it will be a great DX improvement!

A thought: What if user runs `wasp start` and then while developing modifies prisma.schema? Then the warning from us will not arrive, if we are doing the check only on start of `wasp start`, right? Maybe we could also handle that case, by running the check on every compile? The main question is, where does the warning get printed -> I guess it gets printed in the terminal of course, but it might get pushed away by other output. If we had a fixed place in the terminal to print warnings / errors we could print it there, but that will come in the future (powerline idea). So maybe just printing it in terminal is good enough for now.

shayneczyzewski commented:
Hey @Martinsos thanks for the question! The problem with us doing the error on compile is if we throw an error there, they cannot actually run the migrate since compile happens before every migrate runs and they will be stuck forever... I actually tried that approach to start with and learned that lesson myself :D

I agree that is a case we should ideally handle though, and we may just be able to put the same check used in Wasp.Generator.Start into Wasp.Cli.Command.Watch to handle that 👍🏻 I will test that out later today. Thanks!

shayneczyzewski commented:
Note that we are not aiming to resolve this Prisma error case though in this PR (nor am I sure exactly how we would solve it atm):

```
Server (stderr): Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
Server (stderr): In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
Server (stderr):     at new PrismaClient (/Users/shayne/dev/wasp/waspc/examples/todoApp/.wasp/out/server/node_modules/.prisma/client/index.js:3:11)
Server (stderr):     at createDbClient (file:///Users/shayne/dev/wasp/waspc/examples/todoApp/.wasp/out/server/src/dbClient.js:8:24)
Server (stderr):     at file:///Users/shayne/dev/wasp/waspc/examples/todoApp/.wasp/out/server/src/dbClient.js:17:18
Server (stderr):     at ModuleJob.run (internal/modules/esm/module_job.js:152:23)
Server (stderr):     at async Loader.import (internal/modules/esm/loader.js:166:24)
Server (stderr):     at async Object.loadESM (internal/process/esm_loader.js:68:5)
Server (stderr): [nodemon] app crashed - waiting for file changes before starting...
```

This can happen in the examples/todoApp, if you `wasp clean` and then `wasp start` you actually still need a `prisma generate` to generate the prisma client it seems. While `wasp db migrate-dev` can do this too, it is not the root cause.

If you go into wasp/waspc/examples/todoApp/.wasp/out/server and run the following, you can see it is a generation issue and not a migration issue:
```
$ npx prisma generate --schema ../db/schema.prisma
Environment variables loaded from .env
Prisma schema loaded from ../db/schema.prisma

added 2 packages, and audited 3 packages in 1s

found 0 vulnerabilities

added 2 packages, and audited 5 packages in 668ms

found 0 vulnerabilities

✔ Installed the @prisma/client and prisma packages in your project

✔ Generated Prisma Client (2.22.1) to ./node_modules/.prisma/client in 105ms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
```
import { PrismaClient } from './node_modules/.prisma/client'
const prisma = new PrismaClient()
```

After this, `wasp start` works. Perhaps we need to check/run `prisma generate` while building the app?

shayneczyzewski commented on waspc/cli/Wasp/Cli/Command/Watch.hs:
```diff
@@ -80,9 +82,11 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
-      compilationResult <- compileIO waspProjectDir outDir
+      compilationResult <- runExceptT $ do
+        ExceptT $ compileIO waspProjectDir outDir
+        checkIfDbSchemaMatchesChecksum outDir
```
You will see a few instances where I mix `ExceptT String IO ()` and `IO (Either String ())` with a `runExceptT` such that I get short circuiting. Is this a standard way of doing these? I did not in many cases want to for all of them into the `ExceptT` model.

shayneczyzewski replied:
There is another use of `checkIfDbSchemaMatchesChecksum` in a `runExceptT` as well.

Martinsos replied:
I checked `checkIfDbSchemaMatchesChecksum` and I don't see it benefiting from being `ExceptT`. So let's imagine it is `Either` -> in that case you could refactor this as `compilationResult <- compileIO waspProjectDir outDir >> checkIfDbSchemaMatchesChecksum outDir`, and that should be it if I am correct? So it seems to me there is no reason to use `ExceptT` here.

shayneczyzewski replied:
Yeah, I think you are correct. I think I was in gungho monad transformer mode for Prisma errors and tried making everything into an `ExceptT` 😅  Will update thanks 👍🏻

shayneczyzewski replied:
Actually, I think we still do need `ExceptT` to short circuit in both cases (Watch.hs and Start.hs). `>>` here is in the context of `IO` still, so it will just execute all of them even if we return a `Left` from any because IO is like the honey badger here, "it don't care". :D

Check this playground out: https://replit.com/@smcnc/ProfuseThistleBlogclient

I can still make them all `IO (Either String ())` again though if that is easier to read and can just use `ExceptT $ ` before each inside a `runExceptT $ do` to get our desired short circuit behavior.

shayneczyzewski replied:
There are no longer changes to Watch.hs in the current approach. The check now happens during Generation.

shayneczyzewski commented on waspc/examples/todoApp/migrations/prisma.schema-checksum:
```diff
@@ -0,0 +1 @@
+00a14d3339ef0e8cb44b04752e59656d41189df28b6ac02e8c574f695a047bbc
```
Adding our checksum file for the sample app. In real apps it will be in their migrations dir as well so we can even warn them on a fresh checkout, before they do `wasp db migrate-dev`, if they change their entity info.

shayneczyzewski replied:
Arg, yeah I think you are right. There was a different error that masked this https://github.com/wasp-lang/wasp/issues/427 on first run which was normally solved by a `wasp db migrate-dev`, but we do need on a fresh checkout to handle that empty db case. I may need to actually draw out a state diagram for all the cases since it's getting harder and harder to wrap my head around all the permutations of source states vs db states 🤯

shayneczyzewski replied:
Looking back through the discussion, the use case Martin and I were hoping to solve here was if they migrated, and either did a fresh checkout or did a clean, to not warn them unnecessarily they need to migrate again. However, with this file in the source tree I don't think we can solve for the actual problem if they need to migrate anymore. I believe I will have to remove this file from the repo and we may just have to warn them on checkout or clean they need to migrate (even if redundant), and this file being present will serve as "proof" they migrated locally. I will confirm with the table though. Sorry for missing that and good catch 👍🏻

shayneczyzewski replied:
Ok, in thinking it through a bit more, I've come up with the following:
- We cannot include the checksum in the repo, as then we have no way to determine if they should migrate on a fresh checkout and empty db. We need to use the checksum as local "proof" they ran `wasp db migrate-dev`, so it should remain in the generated project.
- We cannot, however, warn them if there is no checksum file either, as on a new project or those without any entities, they should not/cannot migrate anyways, so they cannot generate a checksum.
- We can only start warning after they have successfully executed `wasp db migrate-dev` locally at least once (and thus created a checksum in the generated project).
  - This means though that we are missing several cases, including:
    - Need to run migrate on fresh checkout and empty db
    - Making changes to prisma.schema before checksum is generated

Therefore, the only check we can do is if they have a checksum and it differs from the current schema, we warn. Or if they have a checksum and no schema they are in some crazy state and warn to clean. Still an improvement but I cannot see how we can detect any other nice to have cases.

Thankfully, this is how the checking logic already works now, so after removing checksum from migrations/repo it should behave as above. Does this reflect your thinking @Martinsos @matijaSos? Am I missing anything else? Thanks

shayneczyzewski replied:
I think I would recommend we do in fact remove the checksum from the repo, even it won't help us solve for the fresh checkout and empty db case, to cement the fact that it is a local artifact to help with local migrate decision making.

Martinsos replied:
Great job @shayneczyzewski and @matijaSos  for catching this one, I also totally missed it!

Ok, so I read through what you wrote @shayneczyzewski and I also tried reasoning about it on my own.

I agree that we shouldn't store the checksum in the source project -> while the checksum does tell us that those migrations were generated with the specific prisma schema, we don't know if they were applied on the local database, they might have been applied to a local database by developer A and then checked-out by developer B, and therefore developer B does need to run `wasp db migrate-dev`. However, we can store the checksum in the generated project, therefore locally, in order to track for which version of prisma.schema `wasp db migrate-dev` was called last time locally.

In the case of `wasp clean` or new checkout, there is no generated project, therefore no migrations dir in the generated project and no checksum in the generated project. In that case, why not warn user that they should do migrate-dev? You said because we don't know if they have any entities or not -> if they don't they shouldn't have to run migrate-dev. That makes sense -> but I guess we can check if they have any entities or not and act accordingly, can't we? So if they have any entities, we could warn them that they should run `wasp db migrate-dev`, if migrations directory or checksum file is missing, but if they don't we wouldn't warn them.

Regarding `prisma generate` -> hm this is interesting. So `prisma generate` generates JS/TS bindings for prisma, maybe also something else like parts or whole CLI/Studio and similar. Anyway, it generates prisma artifacts/code, all based on the schema.prisma.
Now, it seems to me that with the warning system we described above, we are covered -> we are warning them to run `wasp db migrate-dev`, which will also do generate, so that is good, all situations are covered -> if they change schema.prisma, we suggest they run it, and if we are not sure if they run it (missing checksum), we again ask them to run it. Except if there are no entities, then we don't ask them anything which is also ok.

However, there is one thought I had -> during development, you might be making changes to schema.prisma -> adding lines, models, ... . Also, modifying some other code based on that, to keep up with the changes. You might not want to run the `migrate` immediately, since you don't want to create a ton of migrations as you are working on it, you first want to get it to compile correctly (if you are using TypeScript) and pass some unit tests. In sucn case, we could be extra smart and run `prisma generate` on every change to schema.prisma automatically! `migrate` we don't want to run automatically, because we don't want to presume when they are ready to do a migration, as migrations introduce some complexity and can go wrong and require some manual intervention and similar. But, `generate` is just generating code -> nothing to go wrong there, and no reason why they wouldn't want that to happen instantly on saving of schema.prisma, right? So we could make the running of `prisma generate` a part of the Wasp.Db.Generator -> when ever we are compiling Wasp code, we would also call prisma generate, if schema.prisma changed since the last compile or is missing (and all of this we would do only if there are entities declared in Wasp).

This said, the last thing with `prisma generate` is an improvement, so we could leave that for a separate issue and for now go with just `wasp db migrate-dev`.

Does this make sense, did I miss something? I tried thinking through all the cases but I might have missed something again.

shayneczyzewski replied:
Hey @Martinsos - thanks for the reply!

Cool, sounds we are in agreement to make it local. Per the warning on fresh checkout/clean:

> That makes sense -> but I guess we can check if they have any entities or not and act accordingly, can't we? So if they have any entities, we could warn them that they should run wasp db migrate-dev, if migrations directory or checksum file is missing, but if they don't we wouldn't warn them.

I agree this would be nice. The only way I can think atm to use as a proxy for "if they have entities" is the existence of the migrations dir. I think that should be a good enough thing to check though. So in this case it will be "if a wasp migrations dir exists, but not a generated project checksum, warn they should migrate". I think that will work 👍🏻 

> But, generate is just generating code -> nothing to go wrong there, and no reason why they wouldn't want that to happen instantly on saving of schema.prisma, right? So we could make the running of prisma generate a part of the Wasp.Db.Generator -> when ever we are compiling Wasp code, we would also call prisma generate, if schema.prisma changed since the last compile or is missing (and all of this we would do only if there are entities declared in Wasp).

I will think on that more. Worst case we can punt it into the new issue I created and do that shortly after this one, but if small and aligned I may be able to sneak it in.

Thanks all!

Martinsos replied:
Can't we read from AppSpec if they have entities?

shayneczyzewski replied:
Unfortunately, I do not think we will have access to an AppSpec at the time we need to do the check. This is currently being called in two places:

1) in CLI watch
```haskell
watch :: Path' Abs (Dir Common.WaspProjectDir) -> Path' Abs (Dir Wasp.Lib.ProjectRootDir) -> IO ()
```

and 2) in Generator start, which is called by CLI start:
```haskell
-- Generator
start :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
```

All the checks have to take place outside of the main Generation/compile step since Generation happens before every migrate, so I could not make it a precondition. So I am making due with only `Path' Abs (Dir ProjectRootDir)`.

Martinsos replied:
Hmmmm ok interesting!

I will try to recap things a bit for my sake:
1. We want to warn use that they need to run `wasp db migrate-dev`.
2. We do that by checking if schema.prisma in the generated code matches the last schema that we used for performing migration -> if not, it means this new schema might need migration, so we show that warning.
3. We store checksum in the generated dir, so it is local, which makes sense.
4. If there is no checksum, we go safe route and show potentially redundant warning that they need to migrate.
5. There is an exception: if user has no entities, then we should not show any warning.

When looking at this, I imagine we could get it working by throwing a warning (not a fatal error) somewhere from inside the compiler -> it could be right after the generation step of compiler but still inside of compiler, or it could be from inside the generator. This way we would have access to AppSpec, and since it is a warning, it won't stop the execution. The question is how do we make it be just a warning though, but the truth is we will likely have other warnings also in the future, so we will probably need to implement support for them anyway at some point.

I am talking a bit abstract while you are actually on the "terrain", so maybe I am missing some practical issues here -> do you think this would be doable, how hard is it?
Could you add additional step into Generator, somewhere in Generator.hs maybe, that checks for this and returns warning, something in that direction, would that work ok?

From what I got you didn't want to make this part of compilation because it would crash the compilation and therefore also prevent us from doing migrate since it calls compilation first. But if it is a warning and doesn't crash anything, then that would not be a problem right? But this makes it impossible for you to access AppSpec, which means we can't make "smart" decisions, like checking there are any entities present in the project.

shayneczyzewski replied:
Yep, your recap sounds accurate to me! I agree, the non-fatal warning approach during compilation/generation is ideal. It makes me think pausing on this until generator errors https://github.com/wasp-lang/wasp/issues/123 is done feels like the best strategy. Otherwise, we will likely have to refator this after anyways.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+checkIfDbSchemaMatchesChecksum' Nothing (Just _dbSchemaChecksumFp) =
+  throwError "Your .wasp output directory appears malformed. Please clean by running wasp clean."
+checkIfDbSchemaMatchesChecksum' Nothing Nothing = return ()
+
```
Should all of the above functions live here? I think it's fine, but would appreciate a sanity check. Thanks!

shayneczyzewski commented on waspc/src/Wasp/Generator/Start.hs:
```diff
@@ -5,19 +5,27 @@ where
+    checkIfDbSchemaMatchesChecksum projectDir
+    start' projectDir
+
+start' :: Path' Abs (Dir ProjectRootDir) -> ExceptT String IO ()
```
The main reason I changed this type to `ExceptT String IO ()` and not `compileIO` used in Watch.hs was this was now a helper local to this file and it made it easier to get the short circuiting. `compileIO` was used in lots of areas and it was better to keep that `IO (Either String ())` in my book. But again, I'm unsure where the line is on these things so welcome feedback. Thanks!

shayneczyzewski replied:
And to recap the problem without resorting to `ExceptT`, I have two or more `IO (Either String ())` functions and would like short circuiting behavior.

shayneczyzewski replied:
No longer updating Start.hs, as the check now happens in Generator.

shayneczyzewski commented on waspc/src/Wasp/Util.hs:
```diff
@@ -102,3 +116,20 @@ concatShortPrefixAndText prefix text =
+  deriving (Show, Eq, Ord, Aeson.ToJSON, Aeson.FromJSON)
+
+bytestringToHex :: B.ByteString -> Hex
+bytestringToHex = Hex . concatMap (printf "%02x") . B.unpack
```
Lifted all the above, and the tests, from our in-progress pair programming work.

shayneczyzewski commented on waspc/cli/Wasp/Cli/Common.hs:
```diff
@@ -44,3 +45,6 @@ extCodeDirInWaspProjectDir = [reldir|ext|]
 waspSays what = putStrLn $ Term.applyStyles [Term.Yellow] what
+
+waspScreams :: String -> IO ()
+waspScreams what = putStrLn $ Term.applyStyles [Term.Red] what
```
I figured it would be nice to be able to display errors in CLI in red vs normal yellow, so anytime we to not throw an error from CLI but still want to report an error, we can scream. :)

shayneczyzewski replied:
Yeah I was thinking this would be nice for any CLI cases where we handle/show errors, but do not throw `CommandError`s

shayneczyzewski commented on waspc/cli/Wasp/Cli/Command/Common.hs:
```diff
@@ -55,6 +57,9 @@ findWaspProjectRootDirFromCwd = do
 waspSaysC = liftIO . waspSays
 
+waspScreamsC :: String -> Command ()
+waspScreamsC = liftIO . waspScreams
```
If this is too silly let me know. Says = yellow, Screams = red :)

matijaSos reviewed (commented):
Nice work @shayneczyzewski, the code looks really good and is easy to read! I didn't go deeper in some transformer stuff I am missing a bit of context for rn (letting that to @Martinsos), but tried to focus more on DX.

shayneczyzewski commented on waspc/cli/Wasp/Cli/Command/Common.hs:
```diff
@@ -55,6 +57,9 @@ findWaspProjectRootDirFromCwd = do
 waspSaysC = liftIO . waspSays
 
+waspScreamsC :: String -> Command ()
+waspScreamsC = liftIO . waspScreams
```
Yep, good point. I was just trying to stay consistent with existing `waspSaysC` but agree that is an option too.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -50,6 +55,12 @@ dbSchemaFileInProjectRootDir = dbRootDirInProjectRootDir </> dbSchemaFileInDbRoo
 dbMigrationsDirInDbRootDir = [reldir|migrations|]
 
+dbSchemaChecksumFileInDbMigrationsDir :: Path' (Rel DbMigrationsDir) File'
+dbSchemaChecksumFileInDbMigrationsDir = [relfile|prisma.schema-checksum|]
```
I think internal files being noted as `.wasp` somehow makes sense! 👍🏻

Martinsos replied:
Yeah I like it!
Btw, isnt it `schema.prisma`, and not `prisma.schema`? I myself make this mistake many time and am never sure, but I just checked their docs and I see that they say `schema.prisma`.
I would suggest name `schema.prisma.wasp-checksum`. So it follows that convention of `somefile.txt.old` that I have seen sometimes on Linux. Or `foo.cpp.swp`. Usually it is the same name together with extension, and then suffix is added but starting with a dot.

shayneczyzewski replied:
Oof woops on the swapping, thanks for catching. `schema.prisma.wasp-checksum` sounds good.

Martinsos replied:
Btw I wonder if we should still be putting it in the migrations dir? Now we can put it anywhere, right? Since we don't care about copying it up? Or actually, it would be better to have it outside of migrations dir, so it doesn't by accident get copied up? Hmmm not sure but asking questions.

shayneczyzewski replied:
Yeah, I moved out of migrations dir when we wanted to keep purely local and now it just lives alongside `schema.prisma` as `.wasp/out/db/schema.prisma.wasp-checksum`

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+  dbSchemaExists <- liftIO $ doesFileExist dbSchemaFp
+  let maybeDbSchemaFp = if dbSchemaExists then Just dbSchemaFp else Nothing
+  dbSchemaChecksumExists <- liftIO $ doesFileExist dbSchemaChecksumFp
+  let maybeDbSchemaChecksumFp = if dbSchemaChecksumExists then Just dbSchemaChecksumFp else Nothing
```
Yeah good call. I had it that way and went back and forth between combining with prep/use vs combining with names. Since you noticed too I'll consider a vote in that direction and swap back.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+
+checkIfDbSchemaMatchesChecksum :: Path' Abs (Dir ProjectRootDir) -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum genProjectRootDir = do
+  dbSchemaExists <- liftIO $ doesFileExist dbSchemaFp
+  let maybeDbSchemaFp = if dbSchemaExists then Just dbSchemaFp else Nothing
```
I didn't have a good way either, but open to a 1-liner if anyone can golf it!

Martinsos replied:
Ha yes it feels like it but the problem is we get only Boolean so we have to somehow again add that information about FilePath to it -> I don't see an easy way to golf it actually hm!

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+    then return ()
+    else throwError "Your Prisma schema has changed, you should run wasp db migrate-dev."
+checkIfDbSchemaMatchesChecksum' (Just _dbSchemaFp) Nothing =
+  -- Cannot throw the error below because they may have no entities in the schema.prisma file.
```
The problem here is both: (1) legacy projects won't have a checksum (but that is minor), but more importantly (2) a brand new project that has no entities will still have a schema, but no checksum (since they don't need to, and indeed cannot migrate). We have no way to know in that case, unless we parse the file, if it "should" have to be migrated (and thus have a checksum) or not without looking at the contents of the schema.prisma. Not sure if we want to go down that path in case that file changes in the future? Feels brittle. So here, we just currently don't issue any error.

I plan to make a state diagram today to validate what is happening right now vs all the permutations, since as you noted with the checksum in the repo, it gets complicated fast. 😅

Martinsos replied:
Can empty schema.prisma file exist? I would expect that in such case, there is no schema.prisma file? But now that I think about it, I have never checked what is the situation in such case.

shayneczyzewski replied:
There should always be a schema.prisma file with some contents, even if they have no entities. Here is what an empty one looks like after a `wasp new` and `wasp start`:

```
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

generator client {
  provider = "prisma-client-js"
  output = "../server/node_modules/.prisma/client"
}
```

Martinsos replied:
Ha interesting I didn't know this!
In such case, should migration be done or not? What happens if `migrate` is called? Migration that removes everything from the db? Hmmm I am not sure how to think about it.

shayneczyzewski replied:
Ha, yeah this was a fun wrinkle! If we call `db mgirate-dev` in this state (with no entities) we get a Prisma error:

```
Performing migration...
Prisma schema loaded from ../db/schema.prisma
Datasource "db": SQLite database "dev.db" at "file:./dev.db"

SQLite database dev.db created at file:./dev.db

Already in sync, no schema change or pending migration was found.

Running generate... (Use --skip-generate to skip the generators)
Error:
You don't have any models defined in your schema.prisma, so nothing will be generated.
You can define a model like this:

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
}

More information in our documentation:
https://pris.ly/d/prisma-schema

Error: Migrate dev failed: Migrate (dev) failed with exit code: 1
```

So in this case, they won't have a checksum file (since migrate cannot succeed), and we should _not_ warn them to migrate since they cannot. Here we can do the "check for entities" approach of looking for a migrations dir (or using AppSpec if we had access in a Warnings world).

shayneczyzewski replied:
Here is a comment above the actual check that covers in detail some of these assumptions and the behavior we have now:

```haskell
-- | This function warns the user to run `wasp db migrate-dev` under certain conditions. But first, a few preliminaries:
--   - This function relies on the latest schema.prisma, and thus should run after the writing of FileDrafts has occurred.
--   - A non-empty schema.prisma should always exist in the generated project dir, even if the user has no entities defined.
--     `wasp new` will create one with a `datasource db {}` and `generator client {}`, for example.
--   - A schema.prisma.wasp-checksum file is created/updated each time `wasp db migrate-dev` is run locally and contains checksum(schema.prisma).
--
-- Given that, there are two cases in which we wish to warn the user to run `wasp db migrate-dev`:
-- (1) If schema.prisma.wasp-checksum exists, but is not equal to checksum(schema.prisma), we know they made changes to schema.prisma and should migrate.
-- (2) If schema.prisma.wasp-checksum does not exist, but the user has entities defined in schema.prisma (and thus, AppSpec).
--     This could imply they have never migrated locally, or that they have but are simply missing their generated project dir.
--     Common scenarios for the second warning include:
--       - After a fresh checkout, or after `wasp clean`; possible false positives in these cases, but for safety, it's still preferable to warn.
--       - When they previously had no entities and just added their first.
checkIfDbSchemaWasUsedForLastMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
```

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
 genMigrationsDir :: Wasp -> Maybe FileDraft
 genMigrationsDir wasp =
-  (getMigrationsDir wasp) >>= \waspMigrationsDir ->
+  getMigrationsDir wasp >>= \waspMigrationsDir ->
```
@matijaSos I also felt I need the parenthesses recently until I accepted that normally binary operators have lower precedence than function application. That said I am still not sure sometimes who has precedence in some situations hm.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum :: Path' Abs (Dir ProjectRootDir) -> ExceptT String IO ()
```
I would make the name more elaborate, since it might not be clear what does the `Checksum` in the name stand for.
So, it could be `checkIfDbSchemaMatchesDbSchemaChecksumFile`. It is long, but I think it tells more correctly what it is about. It could even be smth like `checkIfDbSchemaWasUsedForLastMigration` -> maybe this is even better, since it hides implementation details and more clearly says what we want to use this function for. or smth in this direction, whatever is semantically most correct. I am not sure what makes most sense here, so you decide, I am just throwing ideas.

shayneczyzewski replied:
Yep, I like the second one better myself too, thanks! Will update 👍🏻

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+  throwError "Your .wasp output directory appears malformed. Please clean by running wasp clean."
+checkIfDbSchemaMatchesChecksum' Nothing Nothing = return ()
+
+calculateChecksumStringForFile :: FilePath -> IO String
```
This one feels like it should go into Utils.

shayneczyzewski replied:
Moved over as `checksumFromFilePath :: FilePath -> IO Checksum`

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+  throwError "Your .wasp output directory appears malformed. Please clean by running wasp clean."
+checkIfDbSchemaMatchesChecksum' Nothing Nothing = return ()
+
+calculateChecksumStringForFile :: FilePath -> IO String
+calculateChecksumStringForFile file = do
+  contents <- BS.readFile file
+  let Hex checksum = checksumFromByteString contents
+  return checksum
```
I would move this function to Util and make it return Hex, so it is clear what the String it returns is, and then if the user of this function wants to unwrap that Hex and make it a String or smth else, they can, but I wouldn't do in this function as it means we loose some of the information we have earlier than we need to.

```suggestion
calculateChecksumStringForFile :: FilePath -> IO Hex
calculateChecksumStringForFile file = checksumFromByteString <$> BS.readFile file
```

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+  -- throwError "You are missing the Prisma schema checksum. Please run wasp db migrate-dev."
+  return ()
+checkIfDbSchemaMatchesChecksum' Nothing (Just _dbSchemaChecksumFp) =
+  throwError "Your .wasp output directory appears malformed. Please clean by running wasp clean."
```
I feel like if this situation happens, I am not even sure if they should run `wasp clean` or not hm. I mean it can't hurt, but maybe they don't have to do it? So maybe they deleted all entities from their project, and there is a leftover checksum file. Ok. But then they also have redundant migrations in their source project. They should also now to delete those. Hm. I guess it should be up to us to delete the checksum file if there is no schema.prisma, not to ask them to clean it up.

shayneczyzewski replied:
There should always be a non-empty schema file (even if no entities). I'm honestly not sure how they would even get into this state, but figured "clean" is our universal reset so it was the safest suggestion here.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+
+checkIfDbSchemaMatchesChecksum' :: Maybe FilePath -> Maybe FilePath -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum' (Just dbSchemaFp) (Just dbSchemaChecksumFp) = do
+  dbSchemaFileChecksum <- liftIO $ calculateChecksumStringForFile dbSchemaFp
```
I wonder if instead of receving file paths as maybes, it would be better to just catch possible errors here in this function while trying to read the files. That way we don't have to make assumptions about the external function correctly checking if files exist or not, and it also might simplify our logic, not sure?

shayneczyzewski replied:
Yeah sure I can give it a go 👍🏻

Martinsos replied:
I see what you mean -> I am not sure myself what is nicer, maybe (put intended) give it a try and then we will see what is nicer, as you said.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum' :: Maybe FilePath -> Maybe FilePath -> ExceptT String IO ()
```
Why is this function returning ExcepT? It doesn't seem to me that it benefits from using it, there is no short-circuiting happening, right?

shayneczyzewski replied:
I believe I was going for easier downstream usage where the caller can get easy short circuiting. But as you noted in the other place `>>` may have been enough here, will try that instead.

shayneczyzewski replied:
Referenced in a comment above, but needed to short circuit by caller like here: https://replit.com/@smcnc/ProfuseThistleBlogclient when called in both Start.hs and Watch.hs

Martinsos replied:
Yes I was wrong about that sorry! But still good choice to keep this one just an Either, and then above we can easily wrap it into ExceptT.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum :: Path' Abs (Dir ProjectRootDir) -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum genProjectRootDir = do
```
I wonder if all this checking and maybeing is redundant and we should just catch errors if/when they happen while reading the files -> I also commented about it below.

Martinsos reviewed (changes requested):
@shayneczyzewski ok so this issue turned out to be more complete than we initially thought, but I think it is great that we are now putting proper minds into understanding the workflow around migrating and we can figure out how to provide the best experience for user!

I continued on your great analysis and offered some ideas. I also commented on the current code. All together it seems to me we are pretty close, we just need to flush out some details and confirm some assumptions.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -50,6 +55,12 @@ dbSchemaFileInProjectRootDir = dbRootDirInProjectRootDir </> dbSchemaFileInDbRoo
 dbMigrationsDirInDbRootDir :: Path' (Rel DbRootDir) (Dir DbMigrationsDir)
 dbMigrationsDirInDbRootDir = [reldir|migrations|]
 
+dbSchemaChecksumFileInDbMigrationsDir :: Path' (Rel DbMigrationsDir) File'
```
Good thinking on making it a proper `data` and expanding on comments if I cannot encode into name. Will do, thanks 👍🏻

shayneczyzewski replied:
Added as data now 👍🏻 thanks

Martinsos reviewed (commented):
Responded again to comments! I wish I could offer more constructive advice -> I feel like I am able to point at some parts that could potentially be improved, but I don't have enough knowledge/depth at the moment to confidently point in a very clear direction. I do hope however that you will be more insightful than me and manage to get the best out of it.

shayneczyzewski commented:
> Responded again to comments! I wish I could offer more constructive advice -> I feel like I am able to point at some parts that could potentially be improved, but I don't have enough knowledge/depth at the moment to confidently point in a very clear direction. I do hope however that you will be more insightful than me and manage to get the best out of it.

I appreciate all your help and comments so far @Martinsos. It is making it much better, and we are getting very close. I am leaning toward pivoting over to https://github.com/wasp-lang/wasp/issues/123 and knocking that out first, and then wrapping this up. I hate leaving something at 90% of the way there, but I think the best solution is to leverage non-fatal warnings and make the check part of compilation/generation. Otherwise we will have a non-ideal solution for this (relying on migrations dir and called as start/watch CLI checks) that we would likely want to refactor after anyways.

I'll sync with you both during standup on this. 👍🏻

shayneczyzewski commented:
Hey @Martinsos and @matijaSos It is Groundhog Day 🦫  :D This PR is now back open, cleaned up, reflects your comments that still apply, and is making use of the new `GeneratorWarning` capabilities. This means we no longer have to check for this in Start/Watch in CLI, but can instead check during Generation. This cleaned things up quite a bit. The only slight nag is when they need to migrate, and we warn them, when they actually run `db migrate` the warning scrolls by since the check happens in Generator but that is acceptable I think. It goes away after.

I also used our discussion about what cases to warn to make the checking simpler as well. I will spend a few more hours today testing different scenarios, since it all permutation of states is still non-trivial, but I think it is in good shape for a review whenever you have time. No rush, I'm not blocked on feedback, but just wanted to put it out there again so I don't try to refactor my refactors too long and blow up your alerts Martin 😅 lol. Thanks gang! 🙏🏻

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -73,6 +88,37 @@ deleteGeneratedMigrationsDirIfRedundant spec projectRootDir = do
+    entitiesExist = not . null $ getEntities spec
+
+    warnIf :: Bool -> String -> [GeneratorWarning]
+    warnIf b msg = if b then [GenericGeneratorWarning msg] else []
```
This was the snippet I was referring to during standup. Mainly there to help cleanup another if/else, but pretty close to defining a variant of `warn` like some other libs have done:

```haskell
when' :: (Monad m, Monoid a) => Bool -> m a -> m a
when' b ma = if b then ma else return mempty
```

But wasn't sure if I wanted to go that far, and start adding a ton to Util.hs so I just kept it outside of IO and local to the user. Better styles/approaches?

Martinsos replied:
Oho cool Haskell stuff there!

I have never seen this -> I think what you did is pretty nice and it makes sense to keep it localized for now, and if you realize you require a more general approach, we will probably be smarter regarding how to go about it.

when' is interesting -> although I wonder how it could be used in your situation, due to this combo of Monad and Monoid, I don't think it would be easy to use it directly hm?

shayneczyzewski replied:
Cool thanks, I will leave the helper localized to keep it a bit DRYer. In the `when'` case I could replace something like:

```haskell
else return $ warnIf entitiesExist "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."
```

with:

```haskell
else when' entitiesExist (return [GenericGeneratorWarning "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."])
```

But not really worth adding to Util IMO just yet.

Martinsos commented on waspc/src/Wasp/Generator.hs:
```diff
@@ -54,6 +55,14 @@ genApp spec =
     <++> genDb spec
     <++> genDockerFiles spec
 
+afterWriteChecks :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
```
Maybe we could name it `checkGeneratedCode`? That feels a bit more descriptive on what it does. `afterWriteChecks` tells me the checks are done after the "write", but I am not sure what is being checked -> which is ok if those are super general checks, but I believe they will likely always / for a long time be checks of the generated code, so maybe we can be more precise right now?

Or, we could remove this additional layer / concept to "write checks" and just directly call `DB.warnIfPrismaSchemaChanged` or smth like that.

Not super important though.

Martinsos commented on waspc/src/Wasp/Generator.hs:
```diff
@@ -54,6 +55,14 @@ genApp spec =
+afterWriteChecks :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
+afterWriteChecks spec dstDir = DbGenerator.afterWriteChecks spec dstDir
+
+preCleanup :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ()
```
Nice separating it!

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -6,33 +6,40 @@ module Wasp.Generator.DbGenerator
 data DbTemplatesDir
 
+-- | This file represents the checksum of schema.prisma
+-- at the point at which db migrate-dev is run. It is used
```
```suggestion
-- at the point at which db migrate-dev was last run. It is used
```

Martinsos replied:
While I am writing this, I wonder if we should encode this in name? Should it maybe be called `DbSchemaChecksumOnLastMigrateFile`? Or am I just pushing it too far with this? The reasoning is that when I see `DbSchemaChecksumFile` I don't really know what it is. But maybe that is ok not sure. Then we will need longer name everywhere else which is PITA. Hm.

shayneczyzewski replied:
I can update the comment and see about how the name change looks/feels in use 👍🏻

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -73,6 +88,37 @@ deleteGeneratedMigrationsDirIfRedundant spec projectRootDir = do
   where
     projectMigrationsDirAbsFilePath = SP.fromAbsDir $ projectRootDir </> dbRootDirInProjectRootDir </> dbMigrationsDirInDbRootDir
 
+-- | This function warns the user to run `wasp db migrate-dev` under certain conditions. But first, a few preliminaries:
+--   - This function relies on the latest schema.prisma, and thus should run after the writing of FileDrafts has occurred.
+--   - A non-empty schema.prisma should always exist in the generated project dir, even if the user has no entities defined.
+--     `wasp new` will create one with a `datasource db {}` and `generator client {}`, for example.
+--   - A schema.prisma.wasp-checksum file is created/updated each time `wasp db migrate-dev` is run locally and contains checksum(schema.prisma).
```
```suggestion
-- | Checks if user needs to run `wasp db migrate-dev` due to changes they did in
-- schema.prisma, and if so, returns a warning.
-- When doing this, it looks at schema.prisma in the generated project.
--
-- This function makes following assumptions:
--  - schema.prisma will exist in the generated project even if no Entities were defined.
--    Due to how Prisma itself works, this assumption is currently fulfilled.
--  - schema.prisma.wasp-checksum contains the checksum of the schema.prisma
--    as it was during the last `wasp db migrate-dev`.
```

Why this suggestion:
- Clarifies a bit what is the function doing vs what it is assuming.
- Relaxes last assumption to focus on what schema.prisma.wasp-checksum contains vs how it was generated (very similar though).

Btw I am aware that it is easy for me to propose changes once you write the content :D.

shayneczyzewski replied:
No sweat :D, can update comment 👍🏻

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -73,6 +88,37 @@ deleteGeneratedMigrationsDirIfRedundant spec projectRootDir = do
+--     `wasp new` will create one with a `datasource db {}` and `generator client {}`, for example.
+--   - A schema.prisma.wasp-checksum file is created/updated each time `wasp db migrate-dev` is run locally and contains checksum(schema.prisma).
+--
+-- Given that, there are two cases in which we wish to warn the user to run `wasp db migrate-dev`:
+-- (1) If schema.prisma.wasp-checksum exists, but is not equal to checksum(schema.prisma), we know they made changes to schema.prisma and should migrate.
+-- (2) If schema.prisma.wasp-checksum does not exist, but the user has entities defined in schema.prisma (and thus, AppSpec).
+--     This could imply they have never migrated locally, or that they have but are simply missing their generated project dir.
+--     Common scenarios for the second warning include:
+--       - After a fresh checkout, or after `wasp clean`; possible false positives in these cases, but for safety, it's still preferable to warn.
+--       - When they previously had no entities and just added their first.
```
Very nice!

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -73,6 +88,37 @@ deleteGeneratedMigrationsDirIfRedundant spec projectRootDir = do
+--     Common scenarios for the second warning include:
+--       - After a fresh checkout, or after `wasp clean`; possible false positives in these cases, but for safety, it's still preferable to warn.
+--       - When they previously had no entities and just added their first.
+checkIfDbSchemaWasUsedForLastMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
```
Why not return `Maybe GeneratorWarning`? List fuctions well because `[42]` can be seen as `Just 42` and `[]` can be seen as `Nothing`, but if we know there want be multiple warnings, it would be nicer to use a tighter type.

shayneczyzewski replied:
Yeah good point - I think here we can go tighter, as there probably won't be 2 and can change if we need another 👍🏻

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -73,6 +88,37 @@ deleteGeneratedMigrationsDirIfRedundant spec projectRootDir = do
+--       - After a fresh checkout, or after `wasp clean`; possible false positives in these cases, but for safety, it's still preferable to warn.
+--       - When they previously had no entities and just added their first.
+checkIfDbSchemaWasUsedForLastMigration :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
+checkIfDbSchemaWasUsedForLastMigration spec projectRootDir = do
```
Name is very good, I would maybe rename to `warnIfDbSchemaChangedSinceLastMigration`, since it says what happens after the check -> from the "check" I would expect it to say true/false, but this function actually does one more thing: the check first and then also warns if needed.

shayneczyzewski replied:
I like going from check to warn, sounds good 👍🏻

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -28,7 +32,7 @@ printJobMsgsUntilExitReceived chan = do
     J.JobExit {} -> return ()
 
--- | Migrates in the generated project context and then copies the migrations dir back
+-- | Migrates in the generated project context, stores checksum, and then copies the migrations dir back
```
checksum of what?

Martinsos replied:
you might even consider ommiting it since it is somewhat of an implementation detail.

shayneczyzewski replied:
Prisma schema checksum... but yeah it may be too implementation specific. 🤔 May nix it

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -38,9 +42,14 @@ migrateDevAndCopyToSource dbMigrationsDirInWaspProjectDirAbs genProjectRootDirAb
 
+finalizeMigration :: Path' Abs (Dir ProjectRootDir) -> Path' Abs (Dir DbMigrationsDir) -> IO (Either String ())
+finalizeMigration genProjectRootDirAbs dbMigrationsDirInWaspProjectDirAbs = do
+  writeDbSchemaChecksumToFile genProjectRootDirAbs
```
Maybe the opposite order? Can there be any problem if execution dies for some reason after writing checksum file but before copying migrations back to source? I am not sure, it seems like it shouldn't be a big issue, but still reversing the order sounds a tad bit safer?

shayneczyzewski replied:
Yep np, I can make copy migrations happen first 👍🏻

Martinsos commented on waspc/src/Wasp/Util.hs:
```diff
@@ -12,16 +18,25 @@ module Wasp.Util
     (<++>),
     (<:>),
+    bytestringToHex,
+    Hex (..), -- TODO(shayne): Should we put this into Internal? Or make hexFromString smart constructor?
```
That's a good question. Maybe `hexToString` is a way to go for now! And you would expose `Hex` but not `Hex (..)`.

shayneczyzewski replied:
I just lifted this from our pairing :D but yeah I can do `hexToString` and just expose that and `Hex` for now!

Martinsos commented on waspc/src/Wasp/Util.hs:
```diff
@@ -131,3 +146,25 @@ infixr 5 <:>
+checksumFromByteString = bytestringToHex . SHA256.hash
+
+checksumFromFilePath :: FilePath -> IO Checksum
+checksumFromFilePath file = do
```
You can turn this into a nice one-liner :)!

shayneczyzewski replied:
Ah yeah, can go `checksumFromFilePath file = checksumFromByteString <$> B.readFile file` but I'll spend a few min thinking if I can get point free too :D

Martinsos replied:
Haha oh no you are becoming point-free person :D!
I can imagine doing `(checksumFromByteString <$>) . B.readFile` but that doesn't feel like much improvement I think.

shayneczyzewski replied:
Haha nice, yeah I may just keep file in there but good one!

Martinsos reviewed (commented):
@shayneczyzewski looks great!! This is not much simpler and that is awesome.

I left some comments, nothing major, mostly some naming, rephrasing and small things.

I expected that this warning in CLI will get pushed away, yes, hm. But I guess that mostly happens on first run of `wasp start` and then later it is more visible, when changes are done in wasp source code and `wasp start` re-compiles the project, because there is less additional output after the warning?
Next step would be to actually capture the warning in CLI and show it in some way that ensures it doesn't get lost. For example, we could show it after `npm start` is done with its output (which we can't detect at the moment, but maybe we could implement some heuristics).
Or we could do the "powerline" idea and then show warnings that way.

Anyway, this is great for now!

Martinsos reviewed (approved):
I approved since most of the stuff is done, so you can merge it if you feel you are ready! But I can also check anything additional if you would like me to.

shayneczyzewski commented:
> I approved since most of the stuff is done, so you can merge it if you feel you are ready! But I can also check anything additional if you would like me to.

Thanks @Martinsos! I just wrapped up my commit with updates for your feedback: https://github.com/wasp-lang/wasp/pull/424/commits/6900578c32c35aac6e905335b779f6c9e7869c3a I appreciate all of it. I took nearly all of it. 

The one thing I did leave behind was keeping the notion of `checkGeneratedCode` in the Generator. I worried explicitly calling that db migration check function was mixing the levels of abstraction pretty widely, and do feel we will likely have future checks to add, so the placeholder felt right.

I also left the name `DbSchemaChecksumFile` but updated the comment, as adding the additional context to the name really made it and the others super long.

Otherwise, I think we are good to go, thanks for the eyes and feedback!

Martinsos commented:
> > I approved since most of the stuff is done, so you can merge it if you feel you are ready! But I can also check anything additional if you would like me to.
> 
> Thanks @Martinsos! I just wrapped up my commit with updates for your feedback: [6900578](https://github.com/wasp-lang/wasp/commit/6900578c32c35aac6e905335b779f6c9e7869c3a) I appreciate all of it. I took nearly all of it.
> 
> The one thing I did keep included was the notion of `checkGeneratedCode` in the Generator. I worried explicitly calling that db migration check function was mixing the levels of abstraction pretty widely, and do feel we will likely have future checks to add, so the placeholder felt right.
> 
> I also left the name `DbSchemaChecksumFile` but updated the comment, as adding the additional context to the name really made it and the others super long.
> 
> Otherwise, I think we are good to go, thanks for the eyes and feedback!

Sounds good @shayneczyzewski :)!!!
